runtime/runc_jailer.go

```diff
@@ -0,0 +1,351 @@
+// device nodes, hard links, and/or bind-mount targets
+func (j *runcJailer) BuildJailedRootHandler(cfg *Config, socketPath *string, vmID string) firecracker.Handler {
+	ociBundlePath := j.OCIBundlePath()
+	rootPath := j.RootPath()
+	*socketPath = filepath.Join(rootPath, "api.socket")
+
+	return firecracker.Handler{
+		Name: jailerHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			const mode = os.FileMode(0700)
+			// Create the proper paths needed for the runc jailer
+			j.logger.WithField("rootPath", rootPath).Debug("Creating root drive path")
+			if err := mkdirAndChown(rootPath, mode, j.uid, j.gid); err != nil {
+				return errors.Wrapf(err, "%s failed to mkdirAndChown", rootPath)
+			}
+
+			rootPathToConfig := filepath.Join(ociBundlePath, "config.json")
+			j.logger.WithField("rootPathToConfig", rootPathToConfig).Debug("Copying config")
+			if err := copyFile(runcConfigPath, rootPathToConfig, 0444); err != nil {
+				return errors.Wrapf(err, "failed to copy config from %v to %v", runcConfigPath, rootPathToConfig)
+			}
+
+			j.logger.Debug("Overwritting process args of config")
+			if err := j.overwriteConfig(cfg, filepath.Base(m.Cfg.SocketPath), rootPathToConfig); err != nil {
+				return errors.Wrap(err, "failed to overwrite config.json")
+			}
+
+			// copy the firecracker binary
+			j.logger.WithField("root path", rootPath).Debug("copying firecracker binary")
+			newFirecrackerBinPath := filepath.Join(rootPath, filepath.Base(cfg.FirecrackerBinaryPath))
+			if err := copyFile(
+				cfg.FirecrackerBinaryPath,
+				newFirecrackerBinPath,
+				0700,
+			); err != nil {
+				return errors.Wrapf(err, "could not copy firecracker binary from path %v", cfg.FirecrackerBinaryPath)
+			}
+			if err := os.Chown(newFirecrackerBinPath, int(j.uid), int(j.gid)); err != nil {
+				return errors.Wrap(err, "failed to change ownership of binary")
+			}
+
+			// Link the kernel image
+			newKernelImagePath := filepath.Join(rootPath, kernelImageFileName)
+			j.logger.WithField("newKernelImagePath", newKernelImagePath).Debug("linking kernel image")
+			if err := os.Link(m.Cfg.KernelImagePath, newKernelImagePath); err != nil {
+				return errors.Wrapf(err, "failed to link kernel image from path %v", m.Cfg.KernelImagePath)
+			}
+			m.Cfg.KernelImagePath = kernelImageFileName
+
+			// link drives to new contents path
+			for i, d := range m.Cfg.Drives {
+				drivePath := firecracker.StringValue(d.PathOnHost)
+				fileName := filepath.Base(drivePath)
+				newDrivePath := filepath.Join(rootPath, fileName)
+
+				if err := os.Link(drivePath, newDrivePath); err != nil {
+					return errors.Wrapf(err, "failed to copy drive %v", drivePath)
+				}
+
+				f, err := os.Open(drivePath)
+				if err != nil {
+					return errors.Wrap(err, "failed to open drive file")
+				}
+
+				if internal.IsStubDrive(f) {
+					j.logger.Debug("stub drive detected")
+					if err := os.Chown(newDrivePath, int(j.uid), int(j.gid)); err != nil {
```

Comment:
I'm a bit concerned about Chown()ing the stub drives here. This code is relatively far away from the code that actually creates the drives. If that code is ever changed, there's a good change that whoever changes that code won't realize the downstream impact of their change on this code. Since this code is directly related to how we handle VM isolation, uncoordinated changes to the behavior are likely to have negative effects on critical functionality.

It'd be much safer to create the drives and set the ownership/permission modes of them in the same place.

Reply:
I went ahead and did a little refactoring to allow for chmoding to occur in the stub drive handler. Let me know if that satisfies your concerns!